src/ape/api/providers.py

```diff
@@ -320,3 +320,16 @@ def send_transaction(self, txn: TransactionAPI) -> ReceiptAPI:
         txn_hash = self._web3.eth.send_raw_transaction(txn.encode())
         receipt = self.get_transaction(txn_hash.hex())
         return receipt
+
+
+class UpstreamProvider(ProviderAPI):
+    """
+    A provider that can also be set as another provider's upstream.
+    """
+
+    @property
+    def connection_str(self) -> str:
```

Comment:
We needed a way to have the URI be part of the API, as it wasn't.
However, I didn't want to assume it was always a URI, hence `connection_str`. Thoughts?

Reply:
This is because Hardhat has to provide an HTTP provider connection string to connect upstream? Seems kind of hacky, but I don't see a better way :thinking:

Reply:
Yeah, ultimately, this is the URL value that goes in `npx hardhat node --fork <url>` (or the config file, but it is easier in our case to use the CLI arg for this).